src/lib/Imp.hs

```diff
@@ -923,7 +931,8 @@ toScalarType b = BaseTy b
 fromEmbed :: Subst a => Embed a -> ImpM a
 fromEmbed m = do
   scope <- variableScope
-  let (ans, (_, decls)) = runEmbed m scope
+  let (ans, (scopeDelta, decls)) = runEmbed m scope
+  extend $ mempty { envScope = scopeDelta }
```

Comment:
Ohhhh maybe it's this: the reasoning above is sound, but it assumes that `translateDecl` only generates fresh variables for the binders that were generated in embed. But, if any decl ends up introducing more than a single Imp variable, then we can end up with those variables getting shadowed by the core equations to be translated afterwards. Did I guess it, or was it something else? Can you please add a comment explaining the reasoning here, so that we don't have to recreate all of this in the future?

Reply:
Actually, I think your original reasoning was correct. This was one of the fixes I tried when I was debugging. It didn't fix anything but I left it in by mistake.

Reply:
But doesn't the second comment outline a potential failure case? Maybe we should still keep it?

In any case, even if you think it was not necessary for this particular bug, I would still be curious to hear more about what the bug was.

Reply:
I didn't entirely follow your second comment. But when we're doing our Core->Imp translation, I don't think it's a problem if there are collisions between the names in the Imp delcs we produce and the Core decls we consume. But I could be wrong. It's always so hard to reason about these name conflict questions. I think we should add a type-level mechanism to help us keep track of the different name spaces, as you've been suggesting for a while.